web/src/main/java/com/epam/esm/controller/TagRESTController.java

```diff
@@ -0,0 +1,153 @@
+        int offset = Pagination.offset(page, size);
+        int totalRecords = service.countAllTags();
+        int pages = Pagination.findPages(totalRecords, size);
+        int lastPage = Pagination.findLastPage(pages, size, totalRecords);
+        Link prevLink = linkTo(methodOn(TagRESTController.class).getAllTags(Pagination.findPrevPage(page), size))
+                .withRel("prev");
+        Link nextLink = linkTo(methodOn(TagRESTController.class).getAllTags(Pagination.findNextPage(page, lastPage), size))
+                .withRel("next");
+        List<OnlyTag> models = (service.getAllOnlyTag(size, offset));
+
+        return CollectionModel.of(models, prevLink, nextLink);
+    }
+
+    /**
+     * Get tag by id
+     *
+     * @param id the id
+     * @return tag
+     */
+    @GetMapping("/tags/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public Optional<TagModel> getTag(@PathVariable long id) {
+        Optional<TagModel> model = service.findById(id);
+        allTagsLink(Optional.ofNullable(model).get().orElseThrow());
+        return model;
+    }
+
+    /**
+     * update tag by id
+     *
+     * @param tag the tag
+     * @param id  the id
+     * @return the exposed tag
+     */
+    @PatchMapping("/tags/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public TagModel updateTag(@RequestBody TagModel tag, @PathVariable long id) {
+        service.updateEntity(id, tag);
+        return tag;
+    }
+
+    /**
+     * delete tag by id
+     *
+     * @param id the id
+     * @return string response
+     */
+    @DeleteMapping("/tags/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public String deleteTag(@PathVariable long id) {
+        service.deleteEntity(id);
+        return "message.delete.tag";
+    }
+
+    /**
+     * @param page the page
+     * @param size the size
+     * @return list tag
+     */
+    @GetMapping("/tags/all")
+    @ResponseStatus(HttpStatus.OK)
+    public CollectionModel<TagModel> getAllTags(@RequestParam("page") int page,
+                                                @RequestParam("size") int size) {
+        int offset = Pagination.offset(page, size);
+        int totalRecords = service.countAllTags();
+        int pages = Pagination.findPages(totalRecords, size);
+        int lastPage = Pagination.findLastPage(pages, size, totalRecords);
+        Link prevLink = linkTo(methodOn(TagRESTController.class).getAllTags(Pagination.findPrevPage(page), size))
+                .withRel("prev");
+        Link nextLink = linkTo(methodOn(TagRESTController.class).getAllTags(Pagination.findNextPage(page, lastPage), size))
+                .withRel("next");
+        List<TagModel> models = service.getAllEntity(size, offset);
+
+        return CollectionModel.of(models, prevLink, nextLink);
+    }
+
+    /**
+     * @return the popular tag from the user
+     * with the maximum sum of all order
+     */
+    @GetMapping("/tags/popular")
+    public TagModel getPopularTagWithUser() {
```

Comment:
![изображение](https://user-images.githubusercontent.com/41171038/194885096-ed257d1d-1357-4fa7-a017-b8ebdfbb4aa3.png)
"of a user" means url like "@GetMapping("users/{id}/popular-tag")". As you see it better to place in UserRESTController

Reply:
Also about naming classes convention. If you use abbreviation in names, it's better wo write first letter in upper case, and the rest in lower case. The result would be: "UserRestController".

+ Why do you use Rest in naming? It's a technology, and didn't give you any information. You haven't named classes SpringCertificate or SpringTagService just because you've used Spting, haven't you?

Reply:
Made. Rename controller, learn for the future